### PR TITLE
server/dining migration

### PR DESCRIPTION
**Pull request by judtinzhang**
migrating penn sdk + labs api dining routes

/dining/venues/ - list of venues
/dining/hours/<venue_id>/ - hours for a venue
/dining/weekly_menu/<venue_id>/ - weekly menu for a venue
/dining/daily_menu/<venue_id>/ - daily menu for a venue
/dining/item/<item_id>/ - info for item
/dining/preferences/ - user preferences
/dining/transactions/ - user transactions
/dining/balance/ - user balance updates
/dining/balances - user aggregated balance info
/dining/projection/ - user dining projection

### CODE REVIEW

File: user/migrations/0003_profile_dining_preferences.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# Generated by Django 3.1.7 on 2021-03-14 04:18
```
Comment (ArmaanT) [inline user/migrations/0003_profile_dining_preferences.py:1]: These all should be merged as well

File: laundry/migrations/0002_auto_20210314_1802.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+# Generated by Django 3.1.7 on 2021-03-14 22:02
```
Comment (ArmaanT) [inline laundry/migrations/0002_auto_20210314_1802.py:1]: All the laundry changes should probably go in a different PR for better discoverability

File: user/serializers.py [hunk 3]
```diff
@@ -13,7 +13,7 @@ class Meta:
 class ProfileSerializer(serializers.ModelSerializer):
     class Meta:
         model = Profile
-        fields = ("expected_graduation", "degrees", "laundry_preferences", "dining_preferences")
+        fields = ("expected_graduation", "degrees", "laundry_preferences", "dining_balance")
```
Comment (ArmaanT) [inline user/serializers.py:16]: I'm not sure this will work since there isn't a `dining_balance` field, right?

File: dining/api_wrapper.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        for meal in day["tblDayPart"]:
+            if isinstance(meal["tblStation"], dict):
+                meal["tblStation"] = [meal["tblStation"]]
+            for station in meal["tblStation"]:
+                if isinstance(station["tblItem"], dict):
+                    station["tblItem"] = [station["tblItem"]]
+    return data
+
+
+def get_meals(v2_response, building_id):
```
Comment (ArmaanT) [inline dining/api_wrapper.py:62]: Is building_id the same as a Venue's ID? If so we could switch that to take in a venue directly

File: dining/views.py [hunk 5]
```diff
@@ -16,6 +14,477 @@
...
+    "MENUS": V2_BASE_URL + "menus&cafe=",
+    "ITEMS": V2_BASE_URL + "items&item=",
+}
+
+VENUE_NAMES = {
+    "593": "1920 Commons",
+    "636": "Hill House",
+    "637": "Kings Court English House",
+    "638": "Kosher Dining at Falk",
+}
```
Comment (ArmaanT) [inline dining/views.py:57]: These constants should probably go in the API wrapper file

**Review by ArmaanT — CHANGES_REQUESTED** [review]
Haven't looked too closely at the views yet, but the structure looks pretty good so far

File: dining/models.py [hunk 6]
```diff
@@ -1 +1,42 @@
...
+
+class Venue(models.Model):
+    venue_id = models.IntegerField()
+    image_url = models.URLField()
+
+    def __str__(self):
+        return str(self.venue_id)
+
+
+class DiningPreference(models.Model):
```
[the remark this answers is not in the document]
Comment (judtinzhang) [reply dining/models.py:15]: yea this one was a bit weird. in LAS, it looks like they cared about the # of times a particular venue was chosen as a preference (ex. in the dining_preference view, they aggregate the total number of times somebody "preferred" 1920's or something). I don't think many-to-many can let you "prefer" something more than once, so I think i'll change it over to a foreign key for now and then will swap it once we get the chance to remodel the dining stuff?

File: dining/api_wrapper.py (hunk 4 above)
Comment (judtinzhang) [reply dining/api_wrapper.py:62]: oh woops i'll rename to venue_id for clarity. but i think it should be okay to just pass in the venue_id just because i'm taking the venue_id directly from the URL for that particular route.

also, speaking of: i have the dining api token / password on my computer, but idk how to put it on the api_wrapper file without just writing it flat out for everyone to see. right now i just put some filler text under where i put the request headers, but do you know how to pre-set them to the token / password without making it public on github?

File: dining/migrations/0010_auto_20210314_1802.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
+# Generated by Django 3.1.7 on 2021-03-14 22:02
```
[the remark this answers is not in the document]
Comment (judtinzhang) [reply dining/migrations/0010_auto_20210314_1802.py:1]: hey, so i ran the command `python manage.py makemigrations --merge dining` but i don't think anything happened? it just said no conflicts detected. was it supposed to do something differently?

[the remark this answers is not in the document]
Comment (judtinzhang) [reply dining/migrations/0010_auto_20210314_1802.py:1]: wait nvm i got it `python manage.py squashmigrations` lol. if i run this can i delete all past migrations except for the initial one?

File: dining/models.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ArmaanT) [reply dining/models.py:15]: Can you talk to the ios/android leads and see if they know what's going on with that? Even if that's the case, we're probably better-off modeling this using a through model in a many2many

File: dining/views.py [hunk 8]
```diff
@@ -16,6 +14,462 @@
 from studentlife.utils import date_iso_eastern
 
 
+# from pytz import timezone
+
+
+"""
```
Comment (ArmaanT) [inline dining/views.py:20]: these comments should be cleaned up

File: dining/views.py [hunk 9]
```diff
@@ -16,6 +14,462 @@
...
+            preference["venue_id"] = venue.venue_id
+
+        return Response({"preferences": aggregated_preferences})
+
+    def post(self, request):
+
+        profile = request.user.profile
+
+        # clears all previous preferences associated with the profile
+        preferences = DiningPreference.objects.filter(profile=profile)
```
Comment (ArmaanT) [inline dining/views.py:184]: `DiningPreference.objects.filter(profile=profile).delete()` should do this in one line

File: dining/views.py [hunk 10]
```diff
@@ -16,6 +14,462 @@
...
+        aggregated_preferences = preferences.values("venue_id").annotate(count=Count("venue"))
+
+        # switches venue's auto-generated id with Penn's venue_id
+        for preference in aggregated_preferences:
+            venue = Venue.objects.get(id=preference["venue_id"])
+            preference["venue_id"] = venue.venue_id
+
+        return Response({"preferences": aggregated_preferences})
+
+    def post(self, request):
```
Comment (ArmaanT) [inline dining/views.py:179]: Yeah the logic here really seems a bit weird

File: dining/views.py [hunk 11]
```diff
@@ -16,6 +14,462 @@
...
+        else:
+            return "05:00"
+
+    def get(self, request):
+
+        profile = request.user.profile
+        balance = DiningBalance.objects.filter(profile=profile).order_by("-date").first()
+
+        if balance is not None:
+            timestamp = balance.date.strftime("%Y-%m-%dT%H:%M:%S") + "-{}".format(
```
Comment (ArmaanT) [inline dining/views.py:256]: I know this was a hack in LAS to start with, but I'm pretty sure we don't need the DST fix since django is timezone aware by default

File: tests/dining/test_views.py [hunk 12]
```diff
@@ -0,0 +1,330 @@
+"""
```
Comment (ArmaanT) [inline tests/dining/test_views.py:1]: Is this supposed to be commented out?

**Review by ArmaanT — CHANGES_REQUESTED** [review]

**Review by ArmaanT — CHANGES_REQUESTED** [review]

**Review by ArmaanT — APPROVED** [review]
